test/models/User.spec.ts

```diff
@@ -180,4 +178,94 @@ describe("User", () => {
       expect(profile.connections[0].name).toBe("google-oauth2");
     });
   });
+
+  describe("validate authenctication code", () => {
+    it("should throw a NoCodeError if a user tries to validate a code but no code is stored", async () => {
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "authentication-code":
+              return null;
+          }
+        },
+      });
+
+      try {
+        await caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' });
+        throw new Error('Should throw')
+      } catch (err: any) {
+        if (err.message !== 'No code found') {
+          throw new Error('Should throw NoCodeError')
+        }
+      }
+    });
+
+    it("should throw a InvalidCodeError if a user tries to validate an incorrect code", async () => {
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "authentication-code":
+              return JSON.stringify({ code: "000000" });
+          }
+        },
+      });
+
+      try {
+        await caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' });
+        throw new Error('Should throw')
+      } catch (err: any) {
+        if (err.message !== 'Invalid code') {
+          throw new Error('Should throw InvalidCode')
+        }
+      }
+    });
+
+    it("should throw a AuthenticationCodeExpiredError if a user tries to validate an incorrect code", async () => {
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "authentication-code":
+              return JSON.stringify({ code: "123456", expireAt: 1684757783145 });
+          }
+        },
```

Comment:
this is very cool though. I don't mean for my comments to detract from that :smile: